### ISSUE TITLE
why the src/node_api.h and node_api_types.h head files were moved to the node repository

### ISSUE THREAD

**Issue by oshunter** [nodejs/node-addon-api#855]
The N-API is a good common abstract for native addons and it's common used on some other projects.
Such as the:
https://github.com/jerryscript-project/iotjs/tree/master/include
https://github.com/yodaos-project/ShadowNode/tree/master/include

And the latest Release 3.x.x，the N-API head files are removed to node repository src folder.
Is it possible to keep the head files on the node-addon-api repository for N-API part for other project use and as part of js/C/C++ addon mechanism and more JS engine were supported(V8/QuickJS/JerryScript)?

**Comment by mhdawson** [comment]
@oshunter let us know if what we describe is what you need.

**Comment by mhdawson** [comment]
Possible package name:  `node-api-headers`

**Comment by mjmacleod** [comment]
Thanks for linking me to this.

It would be useful, to me at least though I imagine some others as well, if this package could also contain a .def file (or similar list of symbols that can easily be transformed into a .def file) containing all exported symbols (as per #720) alongside the headers.

This allows for build servers, CI and so on to build an addon without actually having to e.g. have node.exe/electron.exe present to find the symbols from.

Its possible, with some trial/error/hacks to create such a file from the headers, but it would be ideal for one central repository to simply have such a file instead of every project in a similar situation to mine implementing their own build system hacks for this.

**Comment by mhdawson** [comment]
May be a silly question but is a `.def` file cross platform/operating system or does each platform/os have its own equivalent?

**Comment by mjmacleod** [comment]
As far as I understand it `.def` is windows specific.

I'm not 100% sure what the linux/mac/bsd etc. equivalent is however on these platforms it is possible to ignore undefined symbols, so its possible to just do this and link without the executable.

This is of course not 100% ideal because it also ignores other actually undefined symbols in the library, however you do at least get a warning/error about this at runtime, so as long as you test its nothing but a minor headache.
I imagine theres some way to pass `ld` a list of symbols to ignore but I'm not sure what it is...

Anyway, in short a `.def` is windows specific but similar files for other platforms are not strictly necessary, it is at least possible to build for them without them, while windows builds (if e.g. cross compiling or compiling with msys) are not really possible without them.

**Comment by NickNaso** [comment]
Hi everybody,
I had a similar experience trying to create a N-API binding for Go. On the build process at some point the linker will complain about `undefined symbols`. 
On Unix system you could set some flags to ignore `undefined symbols`. In my use case it could be a good choice because N-API is the only C library that I use. 
- Linux: `-Wl,-unresolved-symbols=ignore-all`
- macOS: `-Wl,-undefined,dynamic_lookup`

Anyway on Windows I don't have the possibility to set this kind of flags using `gcc` because  they are not supported so what I did is to create a fake / stub N-API library with all the symbols that I needed. 

From my experience an idea could be a package that provides:
- N-API headers file.
- CLI that help to print the symobls exported by N-API (you can use them on your own build process).
- Maybe expose the API use by the CLI tool if necessary. 

Here I want to share with you my first draft of N-API bindings for Go https://github.com/napi-bindings/go-napi-sys/blob/master/napisys/napisys.go

**Comment by mhdawson** [comment]
From the discussion in the meeting it sounds like a file which just had the list of symbols would have covered the case that @NickNaso had and I think @mjmacleod said the same. 

It should be easy to maintain as it would just be additions when there was a new version of N-API and we believe the list is the same across platforms.

It does mean that the consumers would need to build the specific file they need (for example windows `.def` file from the list but that seems to be ok.

**Comment by mhdawson** [comment]
Next step is to request new repo `nodejs/node-api-headers` following the process (request though admin repo).

**Comment by mhdawson** [comment]
@NickNaso is going to create the request in the admin repo.

**Comment by NickNaso** [comment]
I create the issue now here the ref. https://github.com/nodejs/admin/issues/576

**Comment by oshunter** [comment]
@NickNaso 
Hi, NickNaso
When the node-api-headers repo will be created?
And is this repo will have the release version related with the node release version?

**Comment by oshunter** [comment]
> @oshunter let us know if what we describe is what you need.

Yes, I think it's great to have a separated repo for  the pure headers with release version.

**Comment by NickNaso** [comment]
> @NickNaso
> Hi, NickNaso
> When the node-api-headers repo will be created?
> And is this repo will have the release version related with the node release version?

Hi @oshunter,
usually we need to wait at least 72h after the request for the creation of the repo hoping that tere is no objection.
In my opinion in this repo we need to be related only with the N-API version.

**Comment by oshunter** [comment]
> sually

**Comment by oshunter** [comment]
> > @NickNaso
> > Hi, NickNaso
> > When the node-api-headers repo will be created?
> > And is this repo will have the release version related with the node release version?
> 
> Hi @oshunter,
> usually we need to wait at least 72h after the request for the creation of the repo hoping that tere is no objection.
> In my opinion in this repo we need to be related only with the N-API version.

Thanks！

**Comment by mhdawson** [comment]
@oshunter to set expectations I think we want to do a good amount of review on the first iteration on this as it will be much harder to change later on.  That means it may take us a little while, particularly since many people are on holiday over the next few weeks.

**Comment by NickNaso** [comment]
Hi everybody,
I made a PR https://github.com/nodejs/node-api-headers/pull/1 on node-api-headers repo trying to provide a first implementation. Please if you have time make a review and give me a feedback to go forward.

**Comment by mhdawson** [comment]
Removed stale as the PR above has not yet landed.

**Comment by mhdawson** [comment]
nodejs/node-api-headers#1 landed so closing out.
